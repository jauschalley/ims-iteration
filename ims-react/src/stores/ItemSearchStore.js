import {observable, action, decorate} from "mobx";
import RequestService from '../services/RequestService';
import config from '../config';

class ItemSearchStore {
  searchResults = null;

  searchItems = (formBody) => {
    RequestService.post(`${config.backendUrl}/publications`, formBody, (response) => {
      this.searchResults = response.data.items;
    });
  }
}

decorate(ItemSearchStore, {
  searchResults: observable,
  searchItems: action
});

export default ItemSearchStore;