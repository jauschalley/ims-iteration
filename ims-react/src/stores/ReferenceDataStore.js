import {observable, decorate} from "mobx";

class ReferenceDataStore {
  referenceData = null;

  constructor(referenceData = null) {
    this.referenceData = referenceData || require('../data/lookups.json');
  }
}

decorate(ReferenceDataStore, {
  referenceData: observable,
});

export default ReferenceDataStore;