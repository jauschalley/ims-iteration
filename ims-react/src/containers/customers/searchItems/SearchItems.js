import React from 'react';
import {observer} from "mobx-react";
import {Formik } from 'formik';
import {Row, Container, Col, Form, Button} from 'react-bootstrap';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableSortLabel from '@material-ui/core/TableSortLabel';
import Paper from '@material-ui/core/Paper';
import CwigCard from '../../../components/CwigCard';
import { tablePaginationStore } from '../../../stores/TablePaginationStore';
import TablePaginationGroup from '../../../components/TablePaginationGroup';
import {stableSort} from '../../../components/TableSortGroup';

const buildLink = (url, linkText) => {
  return(
    <a href={url} target="_blank" rel="noopener noreferrer">{linkText}</a>
  );
};

const mapSelectOptions = (options, labelKey, idKey) => (
  options.map((option) => (
    <option value={option[idKey]}>
      {option[labelKey]}
    </option>
  ))
);

const SearchItems = observer(({itemSearchStore, referenceDataStore}) => {
  let refData = referenceDataStore.referenceData;
  let page = tablePaginationStore.page;
  let rowsPerPage = tablePaginationStore.rowsPerPage;

  const [order, setOrder] = React.useState('desc');
  const [orderBy, setOrderBy] = React.useState('SKU');
  const [orderByApiField, setOrderByApiField] = React.useState('calculatedSKU');

  const tableHeadCells = [
    {name: 'SKU', apiFieldName: 'calculatedSKU'},
    {name: 'Title', apiFieldName: 'title'},
    {name: 'Onsite Quantity', apiFieldName: 'qtyOnsite'},
    {name: 'Available Quantity', apiFieldName: 'qtyAvailForOrder'},
    {name: 'PDF', apiFieldName: ''},
    {name: 'HTML', apiFieldName: ''}
  ]

  const handleRequestSort = property => event => {
    const isDesc = orderBy === property && order === "desc";
    setOrder(isDesc ? "asc" : "desc");
    setOrderBy(property);
    setOrderByApiField(tableHeadCells.filter( field => field.name === property)[0].apiFieldName)
  };

  return(
    <Container>
      <Row><h1 className="margin-30">Search Items</h1></Row>
      <Formik
      initialValues={{
        SKU: '',
        Title: '',
        keyword: '',
        typeofinventory: '',
        publicationformat: ''
      }}
      onSubmit={(values) => itemSearchStore.searchItems(values)}
      >
        {formikProps =>
          <Form>
            <CwigCard>
              <Form.Row>
                <Form.Group as={Col} md="4">
                  <Form.Label>SKU</Form.Label>
                  <Form.Control placeholder="SKU" name="SKU" onChange={formikProps.handleChange}></Form.Control>
                </Form.Group>
                <Form.Group as={Col} md="4">
                  <Form.Label>Title</Form.Label>
                  <Form.Control placeholder="Title" name="Title" onChange={formikProps.handleChange}></Form.Control>
                </Form.Group>
                <Form.Group as={Col} md="4">
                  <Form.Label>Keyword</Form.Label>
                  <Form.Control placeholder="Keyword" name="keyword" onChange={formikProps.handleChange}></Form.Control>
                </Form.Group>
              </Form.Row>
              <Form.Row>
                <Form.Group as={Col} md="4">
                  <Form.Label>Inventory Type</Form.Label>
                  <Form.Control as="select" name="typeofinventory" onChange={formikProps.handleChange}>
                    <option>- Select one -</option> 
                    {mapSelectOptions(refData.typesOfInventory, 'productType', 'productTypeID')}
                  </Form.Control>
                </Form.Group>
                <Form.Group as={Col} md="4">
                  <Form.Label>Publication Format</Form.Label>
                  <Form.Control as="select" name="publicationformat" onChange={formikProps.handleChange}>
                    <option>- Select one -</option> 
                    {mapSelectOptions(refData.publicationFormats, 'format', 'formatID')}
                  </Form.Control>
                </Form.Group>
              </Form.Row>
              <Button className="search" onClick={formikProps.handleSubmit}>Search</Button>
            </CwigCard>
          </Form>
        }
      </Formik>
      {itemSearchStore.searchResults && 
      <div>
      <TableContainer component={Paper}>
        <Table aria-label="search customers table">
          <TableHead>
            <TableRow>
              {tableHeadCells.map( cell => (
                cell.name === 'PDF' || cell.name === 'HTML' ?
                  <TableCell key={cell.name}>
                    {cell.name}
                  </TableCell> :
                  <TableCell key={cell.name}>
                    <TableSortLabel
                    active={true}
                    direction={orderBy === cell.name ? order : 'asc'}
                    onClick={handleRequestSort(cell.name)}
                    >
                    {cell.name}
                  </TableSortLabel>
                  </TableCell>
                ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {stableSort(itemSearchStore.searchResults, order, orderByApiField)
              .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
              .map(row => (
                <TableRow key={row.calculatedSKU}>
                  <TableCell>{row.calculatedSKU}</TableCell>
                  <TableCell>{row.title} {row.lastNameName}</TableCell>
                  <TableCell>{row.qtyOnsite}</TableCell>
                  <TableCell>{row.qtyAvailForOrder}</TableCell>
                  <TableCell>{buildLink(row.pubPdfUrl, 'PDF')}</TableCell>
                  <TableCell>{buildLink(row.pubHtmlUrl, 'HTML')}</TableCell>
                </TableRow>
              ))}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePaginationGroup tableRef={itemSearchStore.searchResults}/>
      </div>
      }
    </Container>
  );
});

export default SearchItems;