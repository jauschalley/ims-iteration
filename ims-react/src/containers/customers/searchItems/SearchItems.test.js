import React from 'react';
import renderer from 'react-test-renderer';
import SearchItems from './SearchItems';
import mockSearchResults from './mockSearchResults';
import ItemSearchStore from '../../../stores/ItemSearchStore';
import referenceData from '../newInquiry/mockReferenceData';
import ReferenceDataStore from '../../../stores/ReferenceDataStore';

describe('<SearchItems />', () => {
  it('renders correctly', () => {
    const referenceDataStore = new ReferenceDataStore(referenceData);
    const tree = renderer
      .create(
        <SearchItems 
          itemSearchStore={new ItemSearchStore(mockSearchResults)} 
          referenceDataStore={referenceDataStore} 
        />
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});