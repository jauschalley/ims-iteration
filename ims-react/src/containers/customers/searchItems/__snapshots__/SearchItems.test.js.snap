// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchItems /> renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <h1
      class="margin-30"
    >
      Search Items
    </h1>
  </div>
  <form
    className=""
  >
    <div
      className="cwig-card container"
    >
      <div
        className="row"
      >
        <div
          className="col-md-4"
        >
          <label
            className="form-label"
          >
            SKU
          </label>
          <input
            name="sku"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="sku"
            value=""
          />
        </div>
        <div
          className="col-md-4"
        >
          <label
            className="form-label"
          >
            Title
          </label>
          <input
            name="title"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Title"
            value=""
          />
        </div>
        <div
          className="col-md-4"
        >
          <label
            className="form-label"
          >
            Keyword
          </label>
          <input
            name="keyword"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Keyword"
            value=""
          />
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col-md-4"
        >
          <label
            className="form-label"
          >
            Inventory Type
          </label>
          <select
            name="typeOfInventory"
            onBlur={[Function]}
            onChange={[Function]}
            value=""
          >
            <option
              value={2}
            >
              Administrative Item
            </option>
            <option
              value={1}
            >
              Conference Only Item
            </option>
            <option
              value={3}
            >
              Publication Item
            </option>
          </select>
        </div>
        <div
          className="col-md-4"
        >
          <label
            className="form-label"
          >
            Publication Format
          </label>
          <select
            name="pulbicationFormat"
            onBlur={[Function]}
            onChange={[Function]}
          >
            <option
              value={1}
            >
              Bound
            </option>
            <option
              value={2}
            >
              Unbound
            </option>
            <option
              value={5}
            >
              CD-ROM
            </option>
            <option
              value={6}
            >
              Audio Visual
            </option>
            <option
              value={7}
            >
              Poster
            </option>
          </select>
        </div>
      </div>
      <div
        className="row"
      >
        <button
          className="search btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Search
        </button>
      </div>
    </div>
  </form>
</div>
`;
